src/plotly-graph-card.ts

```diff
@@ -44,10 +43,14 @@ function patchLonelyDatapoints(xs: Datum[], ys: Datum[]) {
   }
 }
 
-function extendLastDatapointToPresent(xs: Datum[], ys: Datum[]) {
+function extendLastDatapointToPresent(
+  xs: Datum[],
+  ys: Datum[],
+  offset: number
+) {
   if (xs.length === 0) return;
   const last = JSON.parse(JSON.stringify(ys[ys.length - 1]));
-  xs.push(new Date());
+  xs.push(new Date(Date.now() + offset));
```

Comment:
`extendLastDatapointToPresent(...)` should be disabled when an offset is set. Those traces reflects values that are not related to the present.

Reply:
Let's discuss this.
An extreme case is a monthly statistic, without extending, it will look like the plot just stops at end of last month.
IMO, of the behavior is to extend to the present, then it should be consistent or it will surprise the user.
I'll put some example screenshots later

Reply:
Well, it depends. 😂
When using offsets, we are talking about things that occurred in the past and is now being shifted.
If we talk about forecast it doesn't make sense to extend, because the forecast is referred to a specific point in time, extending it is nonsensical.
However, when talking a tank level, if the level stays at X liters, and it doesn't change, it would make sense to extend it.

Neither is perfect, but maybe the second scenario is more common so we should stick with your vision. Opinions?

Reply:
True. Give it a try when you test it. I think when you see how a part of the plot disappears as soon as the offset is added you may get the same "this is wrong" feeling.

It is true that it can be seen as a (poor) forecast, but it can also be interpreted as "the last known state at that point".

Let me know what you find out from experimenting. We can always add an option to deactivate the extension

Reply:
For temperatures and statistics (which are currently timed at the start of their period) extending is surprising, because it extends horizontally between one and two hours but the real temperature didn't stay constant. It looks wierd. If you extended by adding the current state instead of the last fetched one it would look better, but that's IMHO too smart, non generalizable and out of scope.

Reply:
Good point. For statistics, extending is particularly bad indeed, irrespective of offsets. (e.g #129).
I think this should be configurable then. 
I'd appreciate opinions on these:
* should `extendToPresent` default to true or false? I'm changing my opinion back and forth.
* from this thread, it seems to me that the pros and cons of extending apply to non offset entities as much as offset ones, and should therefore be considered independent and almost unrelated.

I'm not sure what you mean with:

> extended by adding the current state instead of the last fetched

You mean like what happens with `refresh_interval: auto`?

Reply:
I tend to think `extendToPresent` should default to `false`. In fact, I think it should not exist at all. Extending the data from the last valid value to the end of the configured time range is worthwhile when showing "instantaneous" data like the historical one and thus could default to `true` in that case. Indeed, the sensor actually did keep that state until the end of the time range, or else some `unavailable`, or another value, would have been present.

Note that when using offsets, then on purpose you do not fetch some data that lies between the end of the shown graph and the current time, so extending to the present time *is* harmful because you are going to present false data. The rule IMHO is to extend to the end of the visible timespan, that is to the right of the graph, but a) only if it is prior to the current time b) it should *not* change the actual visible timespan nor ignore any real data that would be in the "end of configured fetch range to current time" zone.

And for data where each point represents a certain time range on its own, then extending it does not work. AFAICT statistics give a new point for new ranges even if the value didn't change so they do not work like the normal history data and thus do not need this workaround anyway. Also, I think HA does not send a "fake" data point in the case of statistics (that fake data point is probably designed exactly to behave like a "extend to the end of the fetched range", BTW).

Reply:
All good points. 
BTW, the fake datapoint is at the start of the range, I didn't see it at the end (or maybe it adds it if the range ends in the past? If that is the case, I have a bug in the code to fix)

I am starting to see how extending to the present (as opposed to the max of the range ) is indeed an issue when offsets come into play. So I think I agree fully with this.

The issue I have with extending by default except if there is an offset, is that there is going to be a jump when you go from offset 0 to anything else.

I'll make a version with all of this configurable so we can play a bit to see if there is a sweet spot.

Maybe something like this:
* default is to extend states (so equal to the standard graph card of HA)
* but not for statistics
* extension is not to the present, but to the limit of the fetched range

I think seeing it in action will make the right solution be evident. I'll code it soon

Reply:
> `extendLastDatapointToPresent(...)` should be disabled when an offset is set. Those traces reflects values that are not related to the present.

@zanna-37 I'm starting to understand why you said this. With offsets, the fetched range doesn't necessarily have anything to do with the present.

Reply:
The fake datapoint at the beginning of the range is probably to "bring into the range" the last datapoint that occured just before the range, so that the left of the graph looks as if it was a cropped version of a bigger graph. I thought that HA did so on the end of the range too, but I may be mistaken.

The "extendLastDatapointToPresent" is not a problem if the offset is moving the trace later. I suppose that what you really want is to extend the last datapoint to the time `now + offset`.

The idea is that when you offset, it is just a translation of the trace by the vector `(offset, 0)`. And the "fake point" of `(now, last_val)` that you add without offset creates a segment from `(last_time, last_val)` to `(now, last_val)` which gets translated into a segment from `(last_time + offset, last_val)` to `(now + offset, last_val)`.

I am not sure if you should crop that `now + offset` to the right end of the graph (to avoid an automatic zoom extending the range in the future initially) or if having data past `now` is not a problem. When the user pans to the right the extension should appear without cropping if offset is positive, though.

Reply:
> The "extendLastDatapointToPresent" is not a problem if the offset is moving the trace later.

If one offsets right (positive) by a month and the visible range is a week, the only the section [now - 1month -1 week, now - 1month] will he fetched] . Extending that to now will just add a huge 1 month straight line (visible when you zoom).
If one offsets left (negative), then it is less problematic

Reply:
> if having data past now is not a problem

X axis scaling is now done explicitly, so this is less of an issue now

Reply:
> Note that my proposition makes having an offset (say a small enough offset that you cannot see, but non-zero nonetheless) totally transparent and not changing suddenly the way the trace looks.

I think i got you now. Whatever we do, it should be equivalent to expanding the true last state first and the offseting. (Not offsetting and then expanding the last fetched one, which may not be the latest state of the sensor)

Reply:
> I think i got you now. Whatever we do, it should be equivalent to expanding the true last state first and the offseting.

If there is an existing state right of the visible range, it should probably be fetched for continuity purposes. Extending the last visible state to the right only works when plotting the values piecewise constant. With other interpolation methods the result is wrong.

If there is no such later state, you may want to assume the state did not change by extending to the right, but never past `now+offset`. I believe this should default to true for normal history and false for statistics, but should be configurable by the user.

Sadly, I don't think HA has a suitable API to fetch « whatever is the next point after this time » so the first paragraph may be infeasible at least for sensors updating very irregularly.

Reply:
> never past now+offset

absolutly

---

>  I believe this should default to true for normal history

Agree

---

> and false for statistics

---

It took me a second to realise why, but I agree now too

---

> but should be configurable by the user.

Also agree. Luckily very easy to implement with the new better structured code.

---

> I don't think HA has a suitable API to fetch « whatever is the next point after this time »

Yeah, I didn't find such an API either. 

---

I'll give this a try now and finally merge this PR